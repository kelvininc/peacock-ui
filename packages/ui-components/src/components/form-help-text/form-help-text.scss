@import '../../assets/styles/globals';

:host {
	/**
	 * @prop --help-text-default-color: Help Text color when state is default.
	 * @prop --help-text-error-color: Help Text color when state is invalid.
	 */

	--help-text-default-color: #{kv-color('neutral-4')};
	--help-text-error-color: #{kv-color('error')};
}

.help-text-container {
	width: 100%;
	display: flex;
	flex-wrap: wrap;
	overflow: hidden;
	margin: $spacing 0 0 $spacing-2x;
	color: var(--help-text-default-color);

	.help-text {
		@include kv-font-label-small-regular;

		margin-right: $spacing;
		color: var(--help-text-default-color);
	}

	span + span::before {
		content: 'â€¢';
		padding-right: $spacing;
	}

	kv-icon {
		--icon-color: var(--help-text-default-color);

		align-items: center;
		margin-right: $spacing-2x;
	}

	&.invalid {
		color: var(--help-text-error-color);

		.help-text {
			color: var(--help-text-error-color);
		}

		kv-icon {
			--icon-color: var(--help-text-error-color);
		}
	}
}
