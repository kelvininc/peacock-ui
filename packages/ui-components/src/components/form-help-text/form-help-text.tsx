import { Component, Host, h, Prop, State, Watch } from '@stencil/core';
import { isEmpty } from 'lodash-es';
import { EValidationState } from '../../types';
import { EIconName } from '../icon/icon.types';

@Component({
	tag: 'kv-form-help-text',
	styleUrl: 'form-help-text.scss',
	shadow: true
})
export class KvFormHelpText {
	/** (optional) Form field help text */
	@Prop({ reflect: true }) helpText: string | string[] = [];
	/** (optional) Form field state */
	@Prop({ reflect: true }) state: EValidationState = EValidationState.None;

	/** Internal help texts state */
	@State() _helpTexts: string[];
	/** Watch the `helpText` property and update internal state accordingly */
	@Watch('helpText')
	helpTextChangeHandler(newValue: string | string[]) {
		this._helpTexts = this.buildHelpTextMessages(newValue);
	}

	componentWillLoad() {
		// Init the states because Watches run only on component updates
		this._helpTexts = this.buildHelpTextMessages(this.helpText);
	}

	private buildHelpTextMessages(value: string | string[]) {
		value = value || [];
		return Array.isArray(value) ? value : [value];
	}

	render() {
		return (
			<Host>
				{!isEmpty(this._helpTexts) && (
					<div class={{ 'help-text-container': true, 'invalid': this.state === EValidationState.Invalid }}>
						{this.state === EValidationState.Invalid && <kv-icon name={EIconName.Error} customClass="icon-16"></kv-icon>}
						{this._helpTexts.map(msg => (
							<span class="help-text">{msg}</span>
						))}
					</div>
				)}
			</Host>
		);
	}
}
