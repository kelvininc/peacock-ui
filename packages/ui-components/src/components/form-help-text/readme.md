# _<kv-form-help-text>_

<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Default -->
<kv-form-help-text helpText="Help text"></kv-form-help-text>

<!-- Default with array of strings-->
<kv-form-help-text [helpText]="helpTexts"></kv-form-help-text>

<!-- Default with error state -->
<kv-form-help-text helpText="Help text" [state]="EValidationState.Invalid"></kv-form-help-text>

```


### React

```tsx
import React from 'react';

import { KvFormHelpText } from '@kelvininc/react-ui-components';

export const FormHelpTextExample: React.FC = () => (
	<>
		{/*-- Default --*/}
		<KvFormHelpText helpText="Help text"/>

		{/*-- Default with array of strings --*/}
		<KvFormHelpText helpText={['Help text 1', 'Help Text2']} />

		{/*-- Default with error state --*/}
		<KvFormHelpText helpText="Help text" state={EValidationState.Invalid} />
	</>
);

```



## Properties

| Property   | Attribute   | Description                     | Type                                                                          | Default                 |
| ---------- | ----------- | ------------------------------- | ----------------------------------------------------------------------------- | ----------------------- |
| `helpText` | `help-text` | (optional) Form field help text | `string \| string[]`                                                          | `[]`                    |
| `state`    | `state`     | (optional) Form field state     | `EValidationState.Invalid \| EValidationState.None \| EValidationState.Valid` | `EValidationState.None` |


## CSS Custom Properties

| Name                        | Description                            |
| --------------------------- | -------------------------------------- |
| `--help-text-default-color` | Help Text color when state is default. |
| `--help-text-error-color`   | Help Text color when state is invalid. |


## Dependencies

### Used by

 - [kv-text-field](../text-field)

### Depends on

- [kv-icon](../icon)

### Graph
```mermaid
graph TD;
  kv-form-help-text --> kv-icon
  kv-text-field --> kv-form-help-text
  style kv-form-help-text fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


