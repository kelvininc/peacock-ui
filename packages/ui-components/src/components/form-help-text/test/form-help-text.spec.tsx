import { SpecPage } from '@stencil/core/internal';
import { KvFormHelpText } from '../form-help-text';
import { newSpecPage } from '@stencil/core/testing';

describe('Form Help Text (unit tests)', () => {
	let page: SpecPage;

	describe('when uses default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvFormHelpText],
				html: `<kv-form-help-text></kv-form-help-text>`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when has a help text', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvFormHelpText],
				html: `<kv-form-help-text helpText="Text Field"></kv-form-help-text>`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when has a help text and state is invalid', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvFormHelpText],
				html: `<kv-form-help-text help-text="Help Text" state="invalid"></kv-form-help-text>`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});
});
