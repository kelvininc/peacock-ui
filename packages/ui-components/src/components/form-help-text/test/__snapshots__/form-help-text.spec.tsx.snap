// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Help Text (unit tests) when has a help text and state is invalid should match the snapshot 1`] = `
<kv-form-help-text help-text="Help Text" state="invalid">
  <mock:shadow-root>
    <div class="help-text-container invalid">
      <kv-icon customclass="icon-16" name="kv-error"></kv-icon>
      <span class="help-text">
        Help Text
      </span>
    </div>
  </mock:shadow-root>
</kv-form-help-text>
`;

exports[`Form Help Text (unit tests) when has a help text should match the snapshot 1`] = `
<kv-form-help-text helptext="Text Field" state="none">
  <mock:shadow-root></mock:shadow-root>
</kv-form-help-text>
`;

exports[`Form Help Text (unit tests) when uses default props should match the snapshot 1`] = `
<kv-form-help-text state="none">
  <mock:shadow-root></mock:shadow-root>
</kv-form-help-text>
`;
